chain/client/src/view_client_actor.rs

```diff
@@ -480,14 +480,25 @@ impl ViewClientActorInner {
                 .flat_map(|outcome| &outcome.outcome.receipt_ids),
         );
 
-        // refund receipt == last receipt in outcome.receipt_ids
-        let mut awaiting_non_refund_receipt_ids: HashSet<&CryptoHash> =
-            HashSet::from_iter(&execution_outcome.transaction_outcome.outcome.receipt_ids);
-        awaiting_non_refund_receipt_ids.extend(execution_outcome.receipts_outcome.iter().flat_map(
-            |outcome| {
-                outcome.outcome.receipt_ids.split_last().map(|(_, ids)| ids).unwrap_or_else(|| &[])
-            },
-        ));
+        // hacky optimization for end-user TX latency: If the last outgoing
+        // receipt of an outcome is a refund receipt, we don't wait for that and
+        // consider the transaction executed or finalized a block earlier.
+        let mut awaiting_non_refund_receipt_ids = awaiting_receipt_ids.clone();
+        for outcome in &execution_outcome.receipts_outcome {
+            let receipt_ids = outcome.outcome.receipt_ids.as_slice();
+            let Some(last_receipt_id) = receipt_ids.last() else {
+                continue;
+            };
+            // need the full receipt to determine if this is a refund or not
+            let Ok(Some(last_receipt)) = self.chain.chain_store.get_receipt(last_receipt_id) else {
```

Comment:
Tiny concern as I see this comment for `get_receipt` function.
```
    /// Fetch a receipt by id, if it is stored in the store.
    ///
    /// Note that not _all_ receipts are persisted. Some receipts are ephemeral,
    /// get processed immediately after creation and don't even get to the
    /// database.
    pub fn get_receipt(&self, receipt_id: &CryptoHash) -> Result<Option<Arc<Receipt>>, Error> {
        self.store.get_ser(DBCol::Receipts, receipt_id.as_ref()).map_err(|e| e.into())
    }
```
Are we sure that the receipts we are trying to fetch would be in the DB? Any any receipt experts chime in and say which receips are considerd ephemeral? cc @nagisa

Reply:
I believe only local receipts are ephemeral. Any receipt that is outgoing cannot be a local receipt because one condition for local receipts is to be produce from the transaction directly.

Reply:
Regarding the DB lookups, in the latest commit we even look up the vector of all outgoing receipts rather than just a single receipt.

This should be fine since we only look it up for receipt id that have not been executed, yet. (I added that check in the latest commit, too). This means, we only read this between when the RPC node has applied a chunk but the chain hasn't validated it yet. Since this is data we only just wrote to the DB, it should be relatively cheap to load. But it does add some computational complexity.

I will create a follow-up issue once this is merged for good.